PurchasesCoreSwift/Public/Error Handling/ErrorUtils.swift

```diff
@@ -0,0 +1,293 @@
+        }
+        let underlyingError = backendUnderlyingError(backendCode: backendCode, backendMessage: backendMessage)
+
+        return error(with: errorCode,
+                     message: errorCode.description,
+                     underlyingError: underlyingError,
+                     extraUserInfo: extraUserInfo)
+
+    }
+
+}
+
+private extension ErrorUtils {
+
+    static func addUserInfo(userInfo: [String: String], error: Error) -> Error {
+
+        let nsError = error as NSError
+        let nsErrorWithUserInfo = NSError(domain: nsError.domain, code: nsError.code, userInfo: userInfo)
+        return nsErrorWithUserInfo as Error
+
+    }
+
+    static func error(with code: ErrorCode,
+                      message: String? = nil,
+                      underlyingError: Error? = nil,
+                      extraUserInfo: [NSError.UserInfoKey: Any]? = nil) -> Error {
+        var userInfo = extraUserInfo ?? [:]
+        userInfo[NSLocalizedDescriptionKey as NSError.UserInfoKey] = message ?? code.description
+        if let maybeUnderlyingError = underlyingError {
+            userInfo[NSUnderlyingErrorKey as NSError.UserInfoKey] = maybeUnderlyingError
+        }
+        userInfo[ErrorDetails.readableErrorCodeKey] = code.codeName
+
+        switch code {
+        case .networkError,
+             .unknownError,
+             .receiptAlreadyInUseError,
+             .unexpectedBackendResponseError,
+             .invalidReceiptError,
+             .invalidAppUserIdError,
+             .operationAlreadyInProgressError,
+             .unknownBackendError,
+             .invalidSubscriberAttributesError,
+             .logOutAnonymousUserError:
+            Logger.error(code.description)
+        case .purchaseCancelledError,
+             .storeProblemError,
+             .purchaseNotAllowedError,
+             .purchaseInvalidError,
+             .productNotAvailableForPurchaseError,
+             .productAlreadyPurchasedError,
+             .missingReceiptFileError,
+             .invalidCredentialsError,
+             .invalidAppleSubscriptionKeyError,
+             .ineligibleError,
+             .insufficientPermissionsError,
+             .paymentPendingError:
+            Logger.appleError(code.description)
+        default:
+            break
+        }
+        let nsError = code as NSError
+        let nsErrorWithUserInfo = NSError(domain: nsError.domain, code: nsError.code,
+                userInfo: userInfo as [String: Any])
+        return nsErrorWithUserInfo as Error
+
+    }
+
+    static func backendUnderlyingError(backendCode: NSNumber?, backendMessage: String?) -> Error {
+
+        let error: Error
+        if let maybeBackendCode = backendCode,
+           // NOTE: Behavior change here, before we would make an error with the code that
+           // came from the backend. Now we create a BackendErrorCode or an Unknown
+           let backendError = BackendErrorCode.init(rawValue: maybeBackendCode.intValue) {
```

Comment:
Behavior change here, before we would make an NSError with the code that came in the JSON response from the backend. Now we create a BackendErrorCode or an BackendErrorCode.unknownError instead. Open to suggestions here.